lib/vsc/filesystem/quota/ktools.py

```diff
@@ -0,0 +1,407 @@
+from vsc.reporting.xdmod.consumer import ConsumerCLI
+
+from vsc.accountpage.client import AccountpageClient
+from vsc.config.base import (
+    GENT, STORAGE_SHARED_SUFFIX, VO_PREFIX_BY_SITE, VO_SHARED_PREFIX_BY_SITE,
+    VSC, VscStorage
+)
+from vsc.utils.mail import VscMail
+
+
+UsageInformation = namedtuple('UsageInformation', [
+    'filesystem', # filesystem name
+    'fileset', # fileset (gpfs)
+    'entity', # the user or VO owning the usage
+    'kind', # the kind of usage info (USR, FILESET, GROUP)
+    'block_usage',  # used quota in KiB
+    'block_soft',  # soft quota limit in KiB
+    'block_hard',  # hard quota limit in KiB
+    'block_doubt',  # the KiB GPFS is not sure about
+    'block_expired',  # tuple (boolean, grace period expressed in seconds)
+    'files_usage',  # used number of inodes
+    'files_soft',  # soft limit for inodes
+    'files_hard',  # hard limit for inodes
+    'files_doubt',  # the inodes GPFS is not sure about
+    'files_expired',  # tuple (boolean, grace period expressed in seconds)
+])
+
+GPFS_GRACE_REGEX = re.compile(
+    r"(?P<days>\d+)\s*days?|(?P<hours>\d+)\s*hours?|(?P<minutes>\d+)\s*minutes?|(?P<expired>expired)"
+)
+
+GPFS_NOGRACE_REGEX = re.compile(r"none", re.I)
+
+QUOTA_USER_KIND = 'USR'
+QUOTA_VO_KIND = 'FILESET'
+
+
+class QuotaException(Exception):
+    pass
+
+
+class DjangoPusher(object):
+    """Context manager for pushing stuff to django"""
+
+    def __init__(self, storage_name, client, kind, dry_run):
+        self.storage_name = storage_name
+        self.storage_name_shared = storage_name + STORAGE_SHARED_SUFFIX
+        self.client = client
+        self.kind = kind
+        self.dry_run = dry_run
+
+        self.count = {
+            self.storage_name: 0,
+            self.storage_name_shared: 0
+        }
+
+        self.payload = {
+            self.storage_name: [],
+            self.storage_name_shared: []
+        }
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        if self.payload[self.storage_name]:
+            self._push(self.storage_name, self.payload[self.storage_name])
+        if self.payload[self.storage_name_shared]:
+            self._push(self.storage_name_shared, self.payload[self.storage_name_shared])
+
+        if exc_type is not None:
+            logging.error("Received exception %s in DjangoPusher: %s", exc_type, exc_value)
+            return False
+
+        return True
+
+    def push(self, storage_name, payload):
+        if storage_name not in self.payload:
+            logging.error("Can not add payload for unknown storage: %s vs %s", storage_name, self.storage_name)
+            return
+
+        self.payload[storage_name].append(payload)
+        self.count[storage_name] += 1
+
+        if self.count[storage_name] > 100:
+            self._push(storage_name, self.payload[storage_name])
+            self.count[storage_name] = 0
+            self.payload[storage_name] = []
```

Comment:
I'm a bit confused by this. not by what it does, but why.

Reply:
To avoid pushing everything at once, so this is a reset.